### PR TITLE
Add Fault Proof withdrawal logic

### LINKED ISSUE #15: Fault Proofs Support
The API for doing a withdrawal is changing with the release of fault proofs. The modifications to the bridge can be found in the [specs](https://github.com/ethereum-optimism/specs/blob/08d3c7189309e9e739d11312ed2ac7c1a43cb582/specs/experimental/fault-proof/stage-one/bridge-integration.md). This tool will need to be updated to support fault proof style withdrawals if it is used in production.

### PR DESCRIPTION
**Pull request by awilliams1-cb**
This PR adds fault proof withdrawal logic, fixing compatibility with `op-mainnet`, `op-sepolia`, and `base-sepolia` networks which now use fault proofs. For right now, non-fault proof withdrawal logic is the default and the feature flag `--fault-proofs` can be used to change the withdrawal logic.

Fixes #15

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>1</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>2</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

**Comment by awilliams1-cb** [comment]
FYI there are some edge cases that the current version of the FP-enabled withdrawer doesn't handle, which includes:

- Creating an output root proposal if one doesn't already exist yet.
- Re-proving withdrawals that were proven against blacklisted dispute games / the game was challenged successfully / the game type changed.

I decided to leave these out in this PR because the diff is already large.

**Comment by cb-heimdall** [comment]
Review Error for cbfyi @ 2024-08-07 10:13:33 UTC
User must have write permissions to review

File: main.go [hunk 1]
```diff
@@ -86,7 +101,19 @@ func main() {
...
+	// check for non-compatible networks with given flags
+	if faultProofs {
+		if n.faultProofs == false {
+			log.Crit("Fault proofs are not supported on this network")
+		}
+	} else {
+		if n.faultProofs == true {
+			log.Crit("Fault proofs are required on this network, please provide the --fault-proofs flag")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (awilliams1-cb) [reply main.go:113]: I added it just so there's a distinction to the user - so for example for testnet the withdrawals aren't instant anymore and I think the `--fault-proofs` flag helps make that clear. But I'm also not attached to the feature flag, so I can also remove it if that's our preference.

**Review by mdehoog — APPROVED** [review]
thank you!

**Comment by Jircs1** [comment]
great work!

**Pull request merged**